teps/0121-refine-retries-for-taskruns-and-customruns.md

```diff
@@ -0,0 +1,201 @@
+---
+status: proposed
+title: Refine Retries for TaskRuns and CustomRuns
+creation-date: '2022-09-08'
+last-updated: '2022-10-14'
+authors:
+- '@XinruZhang'
+- '@jerop'
+- '@lbernick'
+- '@pritidesai'
+see-also:
+- TEP-0069
+---
+
+# TEP-0121: Refine Retries for TaskRuns and CustomRuns
+<!-- toc -->
+- [Summary](#summary)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+- [Related Work](#related-work)
+- [Options Under Consideration](#options-under-consideration)
+  - [Option 1: Implement retries for Standalone TaskRun](#option-1-implement-retries-for-standalone-taskrun)
+  - [Option 2: Implement retries in PipelineRun](#option-2-implement-retries-in-pipelinerun)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Two distinct imperfections on `Retries` we'd like to address in this TEP:
+- `Retries` on `Timeout` is designed inconsistently between TaskRun and CustomRun.
+  - For CustomRun, [the document](https://github.com/tektoncd/community/blob/33ca1d5254a405b1d479f2350443f6c7979a0b72/teps/0069-support-retries-for-custom-task-in-a-pipeline.md#proposal) instructs developers to **set `Timeout` for all retry attempts**. While in the actual implementation, it is **set for each retry attempt**. See the [ref](https://github.com/tektoncd/pipeline/issues/5582).
+  - For TaskRun created out for a PipelineTask, the `Timeout` is **set for each retry attempt**. 
+  - For Standalone TaskRun, there's no `Retries` implemented.
+- Both `PipelineRun` reconciler and `TaskRun`|`CustomRun` reconciler are partially responsible for implementing the `Retries` as of today. See https://github.com/tektoncd/pipeline/issues/5248.
+
+
+### Goals
+1. All four types of runtimes, TaskRuns (independent TaskRun and TaskRun part of a Pipeline) and CustomRuns (independent CustomRun and CustomRun part of a Pipeline) supports retry in which the `TimeOut` (if specified) is set for **each attempt** including no timeout (timeout set to 0).
```

Comment:
this reads more like a proposal than a goal-- this paragraph sounds like we have decided to support retries for all of these objects independently but this is still under discussion.

Reply:
Thanks for pointing this out Lee, right, rephrased to clarify the goal -- "Timeout" is for each retry attempt.